InvalidTypeIdException thrown after kork version bump

Starting with kork version [7.35.0], keel regularly throws the following exception if run against the prestaging database:

[7.35.0]: https://github.com/spinnaker/kork/compare/v7.34.0...v7.35.0

```
com.fasterxml.jackson.databind.exc.InvalidTypeIdException:
Could not resolve type id 'depends-on' as a subtype of 
`com.netflix.spinnaker.keel.api.Constraint`: known type ids = [] at 
[Source: (String)"[
  {"type": "depends-on", "environment": "pretest"},
  {"type": "manual-judgement", "timeout": "PT168H"}
 ]"; line: 1, column: 11]
(through reference chain: java.util.HashSet[0])
```

<details>
<summary>Complete stack trace</summary>

```
org.jooq.exception.DataAccessException: Rollback caused
	at org.jooq.impl.DefaultDSLContext.lambda$transactionResult0$0(DefaultDSLContext.java:578)
	at org.jooq.impl.Tools$12$1.block(Tools.java:4843)
	at java.util.concurrent.ForkJoinPool.managedBlock(ForkJoinPool.java:3323)
	at org.jooq.impl.Tools$12.get(Tools.java:4840)
	at org.jooq.impl.DefaultDSLContext.transactionResult0(DefaultDSLContext.java:588)
	at org.jooq.impl.DefaultDSLContext.transactionResult(DefaultDSLContext.java:505)
	at com.netflix.spinnaker.keel.sql.JooqKt.inTransaction(Jooq.kt:10)
	at com.netflix.spinnaker.keel.sql.SqlDeliveryConfigRepository$itemsDueForCheck$1.invoke(SqlDeliveryConfigRepository.kt:913)
	at com.netflix.spinnaker.keel.sql.SqlDeliveryConfigRepository$itemsDueForCheck$1.invoke(SqlDeliveryConfigRepository.kt:52)
	at com.netflix.spinnaker.keel.sql.SqlRetry$sam$java_util_function_Supplier$0.get(SqlRetry.kt)
	at io.github.resilience4j.retry.Retry.lambda$decorateSupplier$2(Retry.java:177)
	at io.vavr.control.Try.of(Try.java:75)
	at io.vavr.control.Try.ofSupplier(Try.java:92)
	at com.netflix.spinnaker.keel.sql.SqlRetry.withRetry(SqlRetry.kt:24)
	at com.netflix.spinnaker.keel.sql.SqlDeliveryConfigRepository.itemsDueForCheck(SqlDeliveryConfigRepository.kt:912)
	at com.netflix.spinnaker.keel.persistence.CombinedRepository.deliveryConfigsDueForCheck(CombinedRepository.kt:296)
	at com.netflix.spinnaker.keel.persistence.CombinedRepository$$FastClassBySpringCGLIB$$9cfcaa57.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:685)
	at com.netflix.spinnaker.keel.persistence.CombinedRepository$$EnhancerBySpringCGLIB$$df998215.deliveryConfigsDueForCheck(<generated>)
	at com.netflix.spinnaker.keel.actuation.CheckScheduler$checkEnvironments$job$1$1.invokeSuspend(CheckScheduler.kt:110)
	at com.netflix.spinnaker.keel.actuation.CheckScheduler$checkEnvironments$job$1$1.invoke(CheckScheduler.kt)
	at kotlinx.coroutines.intrinsics.UndispatchedKt.startUndispatchedOrReturn(Undispatched.kt:91)
	at kotlinx.coroutines.SupervisorKt.supervisorScope(Supervisor.kt:53)
	at com.netflix.spinnaker.keel.actuation.CheckScheduler$checkEnvironments$job$1.invokeSuspend(CheckScheduler.kt:108)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:738)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
Caused by: com.fasterxml.jackson.databind.exc.InvalidTypeIdException: Could not resolve type id 'depends-on' as a subtype of `com.netflix.spinnaker.keel.api.Constraint`: known type ids = []
 at [Source: (String)"[{"type": "depends-on", "environment": "pretest"}, {"type": "manual-judgement", "timeout": "PT168H"}]"; line: 1, column: 11] (through reference chain: java.util.HashSet[0])
	at com.fasterxml.jackson.databind.exc.InvalidTypeIdException.from(InvalidTypeIdException.java:43)
	at com.fasterxml.jackson.databind.DeserializationContext.invalidTypeIdException(DeserializationContext.java:1761)
	at com.fasterxml.jackson.databind.DeserializationContext.handleUnknownTypeId(DeserializationContext.java:1268)
	at com.fasterxml.jackson.databind.jsontype.impl.TypeDeserializerBase._handleUnknownTypeId(TypeDeserializerBase.java:290)
	at com.fasterxml.jackson.databind.jsontype.impl.TypeDeserializerBase._findDeserializer(TypeDeserializerBase.java:162)
	at com.fasterxml.jackson.databind.jsontype.impl.AsPropertyTypeDeserializer._deserializeTypedForId(AsPropertyTypeDeserializer.java:113)
	at com.fasterxml.jackson.databind.jsontype.impl.AsPropertyTypeDeserializer.deserializeTypedFromObject(AsPropertyTypeDeserializer.java:97)
	at com.fasterxml.jackson.databind.deser.AbstractDeserializer.deserializeWithType(AbstractDeserializer.java:254)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:288)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:245)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:27)
	at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:4218)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3214)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3197)
	at com.netflix.spinnaker.keel.sql.SqlDeliveryConfigRepository$attachDependents$$inlined$let$lambda$1$1.map(SqlDeliveryConfigRepository.kt:995)
	at com.netflix.spinnaker.keel.sql.SqlDeliveryConfigRepository$attachDependents$$inlined$let$lambda$1$1.map(SqlDeliveryConfigRepository.kt:52)
	at org.jooq.impl.ResultImpl.map(ResultImpl.java:1317)
	at org.jooq.impl.AbstractResultQuery.fetch(AbstractResultQuery.java:1564)
	at org.jooq.impl.SelectImpl.fetch(SelectImpl.java:3764)
	at com.netflix.spinnaker.keel.sql.SqlDeliveryConfigRepository$attachDependents$$inlined$let$lambda$1.invoke(SqlDeliveryConfigRepository.kt:371)
	at com.netflix.spinnaker.keel.sql.SqlDeliveryConfigRepository$attachDependents$$inlined$let$lambda$1.invoke(SqlDeliveryConfigRepository.kt:52)
	at com.netflix.spinnaker.keel.sql.SqlRetry$sam$java_util_function_Supplier$0.get(SqlRetry.kt)
	at io.github.resilience4j.retry.Retry.lambda$decorateSupplier$2(Retry.java:177)
	at io.vavr.control.Try.of(Try.java:75)
	at io.vavr.control.Try.ofSupplier(Try.java:92)
	at com.netflix.spinnaker.keel.sql.SqlRetry.withRetry(SqlRetry.kt:35)
	at com.netflix.spinnaker.keel.sql.SqlDeliveryConfigRepository.attachDependents(SqlDeliveryConfigRepository.kt:366)
	at com.netflix.spinnaker.keel.sql.SqlDeliveryConfigRepository.get(SqlDeliveryConfigRepository.kt:349)
	at com.netflix.spinnaker.keel.sql.SqlDeliveryConfigRepository$itemsDueForCheck$1$1$2.map(SqlDeliveryConfigRepository.kt:933)
	at com.netflix.spinnaker.keel.sql.SqlDeliveryConfigRepository$itemsDueForCheck$1$1$2.map(SqlDeliveryConfigRepository.kt:52)
	at org.jooq.impl.ResultImpl.map(ResultImpl.java:1317)
	at com.netflix.spinnaker.keel.sql.SqlDeliveryConfigRepository$itemsDueForCheck$1$1.invoke(SqlDeliveryConfigRepository.kt:932)
	at com.netflix.spinnaker.keel.sql.SqlDeliveryConfigRepository$itemsDueForCheck$1$1.invoke(SqlDeliveryConfigRepository.kt:52)
	at com.netflix.spinnaker.keel.sql.JooqKt$inTransaction$1.run(Jooq.kt:11)
	at org.jooq.impl.DefaultDSLContext.lambda$transactionResult0$0(DefaultDSLContext.java:536)
	... 30 common frames omitted
```

</details>

### Diagnosis: interaction with SpinnakerPluginManager

The SpinnakerPluginManager is instantiating the [KeelApplication] bean too early in the life cycle, leading to the `@PostConstuct` annotated methods of KeelApplication to not be invoked.

Recommended fix by @robzienert is to move all extraneous logic away from the KeelApplication class and onto a different class that is not the main entrypoint.

[KeelApplication]: https://github.com/spinnaker/keel/blob/c2c5500a2f9f6a5d07c6a827f4742f910186525c/keel-web/src/main/kotlin/com/netflix/spinnaker/keel/Main.kt#L61

--- a/keel-web/src/main/kotlin/com/netflix/spinnaker/config/KeelConfigurationFinalizer.kt
+++ b/keel-web/src/main/kotlin/com/netflix/spinnaker/config/KeelConfigurationFinalizer.kt
@@ -1,0 +1,76 @@
+package com.netflix.spinnaker.config
+
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.fasterxml.jackson.databind.jsontype.NamedType
+import com.netflix.spinnaker.keel.actuation.ArtifactHandler
+import com.netflix.spinnaker.keel.api.plugins.ResourceHandler
+import com.netflix.spinnaker.keel.api.plugins.SupportedKind
+import com.netflix.spinnaker.keel.bakery.BaseImageCache
+import com.netflix.spinnaker.keel.constraints.ConstraintEvaluator
+import com.netflix.spinnaker.keel.info.InstanceIdSupplier
+import com.netflix.spinnaker.keel.persistence.ArtifactRepository
+import com.netflix.spinnaker.keel.persistence.DeliveryConfigRepository
+import com.netflix.spinnaker.keel.persistence.ResourceRepository
+import javax.annotation.PostConstruct
+import org.slf4j.LoggerFactory
+import org.springframework.stereotype.Component
+
+/**
+ * Component that wraps up keel configuration once all other beans have been instantiated.
+ */
+@Component
+class KeelConfigurationFinalizer(
+  private val artifactRepository: ArtifactRepository,
+  private val resourceRepository: ResourceRepository,
+  private val deliveryConfigRepository: DeliveryConfigRepository,
+  private val baseImageCache: BaseImageCache? = null,
+  private val instanceIdSupplier: InstanceIdSupplier,
+  private val kinds: List<SupportedKind<*>> = emptyList(),
+  private val resourceHandlers: List<ResourceHandler<*, *>> = emptyList(),
+  private val constraintEvaluators: List<ConstraintEvaluator<*>> = emptyList(),
+  private val artifactHandlers: List<ArtifactHandler> = emptyList(),
+  private val objectMappers: List<ObjectMapper>
+) {
+
+  private val log by lazy { LoggerFactory.getLogger(javaClass) }
+
+  @PostConstruct
+  fun registerResourceSpecSubtypes() {
+    resourceHandlers
+      .map { it.supportedKind }
+      .forEach { (kind, specClass) ->
+        log.info("Registering ResourceSpec sub-type {}: {}", kind, specClass.simpleName)
+        val namedType = NamedType(specClass, kind.toString())
+        objectMappers.forEach { it.registerSubtypes(namedType) }
+      }
+  }
+
+  @PostConstruct
+  fun registerConstraintSubtypes() {
+    constraintEvaluators
+      .map { it.supportedType }
+      .forEach { constraintType ->
+        log.info("Registering Constraint sub-type {}: {}", constraintType.name, constraintType.type.simpleName)
+        val namedType = NamedType(constraintType.type, constraintType.name)
+        objectMappers.forEach { it.registerSubtypes(namedType) }
+      }
+  }
+
+  @PostConstruct
+  fun initialStatus() {
+    sequenceOf(
+        ArtifactRepository::class to artifactRepository.javaClass,
+        ResourceRepository::class to resourceRepository.javaClass,
+        DeliveryConfigRepository::class to deliveryConfigRepository.javaClass,
+        BaseImageCache::class to baseImageCache?.javaClass,
+        InstanceIdSupplier::class to instanceIdSupplier.javaClass
+    )
+        .forEach { (type, implementation) ->
+          log.info("{} implementation: {}", type.simpleName, implementation?.simpleName)
+        }
+
+    log.info("Supporting resource kinds: {}", kinds.joinToString { it.kind.toString() })
+    log.info("Using resource handlers: {}", resourceHandlers.joinToString { it.name })
+    log.info("Using artifact handlers: {}", artifactHandlers.joinToString { it.name })
+  }
+}

--- a/keel-web/src/main/kotlin/com/netflix/spinnaker/keel/Main.kt
+++ b/keel-web/src/main/kotlin/com/netflix/spinnaker/keel/Main.kt
@@ -15,21 +15,7 @@
  */
 package com.netflix.spinnaker.keel
 
-import com.fasterxml.jackson.databind.ObjectMapper
-import com.fasterxml.jackson.databind.jsontype.NamedType
-import com.netflix.spinnaker.keel.actuation.ArtifactHandler
-import com.netflix.spinnaker.keel.api.plugins.ResourceHandler
-import com.netflix.spinnaker.keel.api.plugins.SupportedKind
-import com.netflix.spinnaker.keel.bakery.BaseImageCache
-import com.netflix.spinnaker.keel.constraints.ConstraintEvaluator
-import com.netflix.spinnaker.keel.info.InstanceIdSupplier
-import com.netflix.spinnaker.keel.persistence.ArtifactRepository
-import com.netflix.spinnaker.keel.persistence.DeliveryConfigRepository
-import com.netflix.spinnaker.keel.persistence.ResourceRepository
 import com.netflix.spinnaker.kork.PlatformComponents
-import javax.annotation.PostConstruct
-import org.slf4j.LoggerFactory
-import org.springframework.beans.factory.annotation.Autowired
 import org.springframework.boot.autoconfigure.SpringBootApplication
 import org.springframework.boot.builder.SpringApplicationBuilder
 import org.springframework.context.annotation.Import
@@ -58,80 +44,7 @@ private val DEFAULT_PROPS = mapOf(
 @Import(PlatformComponents::class)
 @EnableAsync
 @EnableScheduling
-class KeelApplication {
-
-  private val log by lazy { LoggerFactory.getLogger(javaClass) }
-
-  @Autowired
-  lateinit var artifactRepository: ArtifactRepository
-
-  @Autowired
-  lateinit var resourceRepository: ResourceRepository
-
-  @Autowired
-  lateinit var deliveryConfigRepository: DeliveryConfigRepository
-
-  @Autowired(required = false)
-  var baseImageCache: BaseImageCache? = null
-
-  @Autowired
-  lateinit var instanceIdSupplier: InstanceIdSupplier
-
-  @Autowired(required = false)
-  var kinds: List<SupportedKind<*>> = emptyList()
-
-  @Autowired(required = false)
-  var resourceHandlers: List<ResourceHandler<*, *>> = emptyList()
-
-  @Autowired(required = false)
-  var constraintEvaluators: List<ConstraintEvaluator<*>> = emptyList()
-
-  @Autowired(required = false)
-  var artifactHandlers: List<ArtifactHandler> = emptyList()
-
-  @Autowired
-  lateinit var objectMappers: List<ObjectMapper>
-
-  @PostConstruct
-  fun registerResourceSpecSubtypes() {
-    resourceHandlers
-      .map { it.supportedKind }
-      .forEach { (kind, specClass) ->
-        log.info("Registering ResourceSpec sub-type {}: {}", kind, specClass.simpleName)
-        val namedType = NamedType(specClass, kind.toString())
-        objectMappers.forEach { it.registerSubtypes(namedType) }
-      }
-  }
-
-  @PostConstruct
-  fun registerConstraintSubtypes() {
-    constraintEvaluators
-      .map { it.supportedType }
-      .forEach { constraintType ->
-        log.info("Registering Constraint sub-type {}: {}", constraintType.name, constraintType.type.simpleName)
-        val namedType = NamedType(constraintType.type, constraintType.name)
-        objectMappers.forEach { it.registerSubtypes(namedType) }
-      }
-  }
-
-  @PostConstruct
-  fun initialStatus() {
-    sequenceOf(
-      ArtifactRepository::class to artifactRepository.javaClass,
-      ResourceRepository::class to resourceRepository.javaClass,
-      DeliveryConfigRepository::class to deliveryConfigRepository.javaClass,
-      BaseImageCache::class to baseImageCache?.javaClass,
-      InstanceIdSupplier::class to instanceIdSupplier.javaClass
-    )
-      .forEach { (type, implementation) ->
-        log.info("{} implementation: {}", type.simpleName, implementation?.simpleName)
-      }
-
-    log.info("Supporting resource kinds: {}", kinds.joinToString { it.kind.toString() })
-    log.info("Using resource handlers: {}", resourceHandlers.joinToString { it.name })
-    log.info("Using artifact handlers: {}", artifactHandlers.joinToString { it.name })
-  }
-}
+class KeelApplication
 
 fun main(vararg args: String) {
   SpringApplicationBuilder()

